lib/rules/prefer-lookaround.ts

```diff
@@ -0,0 +1,625 @@
+    // Reference at the ending position.
+    // e.g.
+    // '$1str$2' -> { ref: 2, range: [5,7] }
+    // '$<foo>str$<bar>' -> { ref: 'bar', range: [9,15] }
+    // '$1str' -> null
+    // '$1str$2str' -> null
+    endRef: ReplaceReference | null
+    // All references including the above.
+    allRefs: ReplaceReference[]
+}
+
+/**
+ * Holds all replacement reference data.
+ *
+ * If the same RegExp instance is used for replacement in 2 places, the number of data in `list` is 2.
+ */
+class ReplaceReferencesList {
+    private readonly list: ReplaceReferences[]
+
+    /** Reference name at the starting position. */
+    public readonly startRefName?: string | number
+
+    /** Reference name at the ending position. */
+    public readonly endRefName?: string | number
+
+    /** All reference names except at the starting position. */
+    public readonly otherThanStartRefNames: Set<string | number>
+
+    /** All reference names except at the starting position. */
+    public readonly otherThanEndRefNames: Set<string | number>
+
+    public constructor(list: ReplaceReferences[]) {
+        this.list = list
+        this.startRefName = list[0].startRef?.ref
+        this.endRefName = list[0].endRef?.ref
+
+        const otherThanStartRefNames = new Set<string | number>()
+        const otherThanEndRefNames = new Set<string | number>()
+        for (const { startRef, endRef, allRefs } of this.list) {
+            for (const ref of allRefs) {
+                if (ref !== startRef) {
+                    otherThanStartRefNames.add(ref.ref)
+                }
+                if (ref !== endRef) {
+                    otherThanEndRefNames.add(ref.ref)
+                }
+            }
+        }
+        this.otherThanStartRefNames = otherThanStartRefNames
+        this.otherThanEndRefNames = otherThanEndRefNames
+    }
+
+    public *[Symbol.iterator](): Iterator<ReplaceReferences> {
+        yield* this.list
+    }
+}
+
+const enum SideEffect {
+    startRef,
+    endRef,
+}
+
+/**
+ * Gets the type of side effect when replacing the capture group for the given element.
+ *
+ * If  the starting and ending capture groups, and the patterns between each have disjoints with each other,
+ * there are no side effects.
+ */
+function getSideEffectsWhenReplacingCapturingGroup(
+    start: CapturingGroup | undefined,
+    others: Element[],
+    end: CapturingGroup | undefined,
+    regexpContext: RegExpContext,
+): Set<SideEffect> {
+    const result = new Set<SideEffect>()
+    if (start && others.length) {
+        if (!hasDisjoint(start, ...others)) {
+            result.add(SideEffect.startRef)
+        }
+    }
+    if (end && others.length) {
+        if (!hasDisjoint(end, ...others)) {
+            result.add(SideEffect.endRef)
+        }
+    }
+    if (start && end) {
+        if (!hasDisjoint(start, end)) {
+            result.add(SideEffect.startRef)
+            result.add(SideEffect.endRef)
+        }
+    }
```

Comment:
Frankly, I am not sure whether this is correct. The method I proposed was intended for the start capturing group.

I _believe_ this could work but I will have to think about this some more. I'll write back later.

Reply:
I thought for a bit and it doesn't work.

```js
> "abaaaba".replace(/ab(a+)/g, "c$1")
'caaaba'
> "abaaaba".replace(/ab(?=a+)/g, "c")
'caaca'
```

As for what I think does work:

A quick reminder before: What property are we trying to verify?
We are trying the verify that the capturing group (CG) behaves in the same way as a lookahead. This is only the case if the end CG does not consume characters that the regex can consume in the next match.

### Without `g` flag
Therefore we can trivially convert the end CG into a lookahead if the `g` flag isn't present because there won't be a "next match".

### With `g` flag
If the `g` flag is present, then this whole thing is more complex. I have 2 conditions: a simple with many false negatives and a complex one.

#### The simple one
We can replace the end CG with a lookahead if the set of all consumed characters of the end CG is disjoint with the first character consumed by the regex.

While this condition guarantees correctness, it's also very conservative and will likely produce false negatives. 

#### The complex one

Now a little diagram. I want to illustrate the possible ways in which the lookahead of the end CG can overlap with the next match. I will call the range asserted by the lookahead of the end CG asserted _E_ and the next match _R_. There are 3 possible configurations in which _E_ and the next match of _R_ can be:

```
1) Disjoint
<~~E~~>
        <--R-->

2) R in E
<~~E~~>
 <-R->

3) E and R overlap but R is not completely in E
<~~E~~>
   <-R->
```

We need case 1 to do the conversion, so we just have to show that cases 2 and 3 are not possible.

Let _X_ be the language accepted by the end CG and let _Y_ be the language accepted by the whole regex. (I use _•_ for concatenation.)

Case 2 is not possible if _X ∩ (Σ* • Y • Σ*) = ∅_.

Let _Z_ be the set of all non-empty prefixes of _Y_. Case 3 is not possible if _X ∩ (Σ* • Z) = ∅_.

Both conditions can be implemented using refa's NFAs.

#### Which one?

Frankly, I don't know whether the performance cost of converting the whole pattern into an NFA is worth it. 

I'll let you decide which one to implement @ota-meshi.

---

Note: Now the conditions for the start and end capturing group are fully independent.

Reply:
Thanks for the detailed explanation!!
I don't understand the `complex one` yet, but I think a simple implementation is probably fine.
I think that even a simple implementation will work pretty well for users who want to use this rule.

Reply:
As I was working, I noticed that the following results were different.

```js
"abaaaba".replace(/(a)ba+/g, "$1c")
// 'acba'
"abaaaba".replace(/(?<=a)ba+/g, "c")
// 'acc'
```

This means that the converted result of `"".replace(/(a)Java/g, '$1foo')` will change.

```js
"aJavaJava".replace(/(a)Java/g, '$1foo')
// 'afooJava'
"aJavaJava".replace(/(?<=a)Java/g, 'foo')
// 'afoofoo'
```

We probably need to check the last element, whether or not the last element is a capturing group.

Reply:
You're right. We need to check that the first character of the start group is disjoint with the last character of the whole pattern.

Reply:
You're right. We need to check that the first character of the start group is disjoint with the last character of the whole pattern.